flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/ShuffleOperator.java

```diff
@@ -0,0 +1,140 @@
+
+import java.io.Serializable;
+import java.util.Map;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.api.java.typeutils.MapTypeInfo;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.operators.coordination.OperatorEventGateway;
+import org.apache.flink.runtime.operators.coordination.OperatorEventHandler;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/**
+ * Shuffle operator can help to improve data clustering based on the key.
+ *
+ * <p>It collects the data statistics information, sends to coordinator and gets the global data
+ * distribution weight from coordinator. Then it will ingest the weight into data stream(wrap by a
+ * class{@link ShuffleRecordWrapper}) and send to partitioner.
+ */
+@Internal
+public class ShuffleOperator<T, K extends Serializable>
+    extends AbstractStreamOperator<ShuffleRecordWrapper<T, K>>
+    implements OneInputStreamOperator<T, ShuffleRecordWrapper<T, K>>, OperatorEventHandler {
+
+  private static final long serialVersionUID = 1L;
+
+  private final KeySelector<T, K> keySelector;
+  // the type of the key to collect data statistics
+  private final TypeInformation<K> keyType;
+  private final OperatorEventGateway operatorEventGateway;
+  // key is generated by applying KeySelector to record
+  // value is the times key occurs
+  private transient Map<K, Long> localDataStatisticsMap;
+  private transient Map<K, Long> globalDataDistributionWeightMap;
+  private transient ListState<Map<K, Long>> globalDataDistributionWeightState;
+
+  public ShuffleOperator(
+      KeySelector<T, K> keySelector,
+      TypeInformation<K> keyType,
+      OperatorEventGateway operatorEventGateway) {
+    this.keySelector = keySelector;
+    this.keyType = keyType;
+    this.operatorEventGateway = operatorEventGateway;
+  }
+
+  @VisibleForTesting
+  ListStateDescriptor<Map<K, Long>> generateGlobalDataDistributionWeightDescriptor() {
+    return new ListStateDescriptor<>(
+        "globalDataDistributionWeight", new MapTypeInfo<>(keyType, TypeInformation.of(Long.class)));
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    localDataStatisticsMap = Maps.newHashMap();
+    globalDataDistributionWeightState =
```

Comment:
The global data distribution should be the same (or samish) for every instance of the operator.
Shall we use `getBroadcastState` instead?
Based on my understanding that would better approximate our requirements for redistributing the state when rescaling. Even if there are some small discrepancies in the values between the different operation instances, we should not be concerned too much about them.

WDYT?

Reply:
@pvary discussed this with me offline. Flink doc said that broadcast is usually used with broadcast stream. But it may just work for regular stream. This is the part yet to be verified. @yegangy0718 can you experiment the broadcast with some unit test (especially for rescale)?

Reply:
My current understanding is, when using broadcast, there will be two streams. One stream will be send to all tasks, and the other stream won’t be send to all tasks. Then connecting them together by calling `connect` . The task can use `getBroadcastState` to get the broadcast stream.
From concept perspective, yes, it’s similar to what we are dong. Global data statistics will be the broadcast one.
But for our case, there is no stream/place publish the broadcast state, where the shuffle operator can get BroadcastState? Do you mean the shuffle operator publishes the data statistics once receiving from coordinator and then in function `initializeState` shuffle operator can get BroadcastState?

Reply:
Let's merge this change, and we can play around the Broadcaststate in a followup PR. WTYT?

Reply:
SGTM. I can create a followup PR to add scaling up test case to check the behavior for BroadcastState.